tools/validate_spec.py

```diff
@@ -79,81 +78,125 @@ def cli(argv: List[str]) -> Union[int, str]:
         return f"output file already exists: {args.json_output}"
 
     try:
-        validation_main(args)
+        identifier = ID(args.message_identifier)
+    except RecordFluxError as e:
+        return f'invalid identifier "{args.message_identifier}" : {e}'
+
+    try:
+        message_value = PyRFLX.from_specs(
+            [str(args.specification)], skip_model_verification=args.no_verification
+        )[str(identifier.parent)][str(identifier.name)]
+    except KeyError:
+        return f'message "{identifier.name}" could not be found in package "{identifier.parent}"'
+
+    except FileNotFoundError as e:
+        return f"specification {e}"
+
+    try:
+        validate(
+            message_value,
+            args.directory_invalid,
+            args.directory_valid,
+            args.json_output,
+            args.abort_on_error,
+        )
     except ValidationError as e:
         return f"{e}"
 
     return 0
 
 
-def validation_main(args: argparse.Namespace) -> None:
-    # pylint: disable=too-many-locals
-    def get_message_to_validate(message_path: Path, is_valid: bool) -> "OriginalMessage":
-        if not message_path.is_file():
-            raise ValidationError(f"{message_path} is not a regular file")
-        message_bytes = message_path.read_bytes()
-        message = OriginalMessage(message_bytes, is_valid, message_path.name)
-        return message
-
-    root_message_id: str = args.message_identifier
-    path_spec: Path = args.specification
-    path_invalid: Optional[Path] = args.directory_invalid
-    path_valid: Optional[Path] = args.directory_valid
-    full_output_path: Optional[Path] = args.json_output
-    abort_on_error: bool = args.abort_on_error
-    no_verification: bool = args.no_verification
-
-    path_generators: Dict[Path, Iterator[Path]] = {}
-    for path in [path_valid, path_invalid]:
-        if path is not None:
-            path_generators[path] = path.glob("*")
+def validate(
+    message_value: MessageValue,
+    directory_invalid: Optional[Path],
+    directory_valid: Optional[Path],
+    json_output: Optional[Path],
+    abort_on_error: bool,
+) -> int:
+
+    with OutputWriter(json_output) as output_writer:
+        for directory_path, is_valid_directory in [
+            (directory_valid, True),
+            (directory_invalid, False),
+        ]:
+            directory = sorted(directory_path.glob("*")) if directory_path is not None else []
+            for path in directory:
+                validation_result = __validate_message(path, is_valid_directory, message_value)
+                output_writer.write_result(validation_result)
+                if not validation_result.validation_success and abort_on_error:
+                    raise ValidationError(f"aborted: message {path} was classified incorrectly")
 
+    return 0
+
+
+def __validate_message(
+    message_path: Path, valid_original_message: bool, message_value: MessageValue
+) -> "ValidationResult":
```

Comment:
Private functions should only have one leading underscore. There is no advantage of using two leading underscores on the module level.

Reply:
Changed in b04365a